src/BlockTemplatesController.php

```diff
@@ -38,6 +38,97 @@ public function __construct() {
 	protected function init() {
 		add_filter( 'get_block_templates', array( $this, 'add_block_templates' ), 10, 3 );
 		add_action( 'template_redirect', array( $this, 'render_block_template' ) );
+		add_filter( 'pre_get_block_template', array( $this, 'maybe_return_blocks_template' ), 10, 3 );
+	}
+
+	/**
+	 * This function checks if there's a blocks template (ultimately it resolves either a saved blocks template from the
+	 * database or a template file in `woo-gutenberg-products/block/templates/block-templates/`)
+	 * to return to pre_get_posts short-circuiting the query in Gutenberg.
+	 *
+	 * @param \WP_Block_Template|null $template Return a block template object to short-circuit the default query,
+	 *                                               or null to allow WP to run its normal queries.
+	 * @param string                  $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param array                   $template_type wp_template or wp_template_part.
+	 *
+	 * @return mixed|\WP_Block_Template|\WP_Error
+	 */
+	public function maybe_return_blocks_template( $template, $id, $template_type ) {
+		$template_name_parts = explode( '//', $id );
+		if ( count( $template_name_parts ) < 2 ) {
+			return $template;
+		}
+		list( , $slug ) = $template_name_parts;
+
+		// Remove the filter at this point because if we don't then this function will infinite loop.
+		remove_filter( 'pre_get_block_template', array( $this, 'maybe_return_blocks_template' ), 10, 3 );
+
+		// Check if the theme has a saved version of this template before falling back to the woo one. Please note how
+		// the slug has not been modified at this point, we're still using the default one passed to this hook.
+		$maybe_template = gutenberg_get_block_template( $id, $template_type );
+		if ( null !== $maybe_template ) {
+			return $maybe_template;
```

Comment:
Shouldn't we add the `pre_get_block_template` filter back again before returning?

Reply:
Great point 👏🏼